tools/perfscale-load-generator/burst/burst.go

```diff
@@ -44,104 +43,116 @@ type BurstJob struct {
 	UUID       string
 	objType    string
 	done       <-chan time.Time
+	watchers   map[string]*watcher.ObjListWatcher
 }
 
 // NewBurstJob
-func newBurstJob(virtClient kubecli.KubevirtClient, workload *config.Workload, d <-chan time.Time) *BurstJob {
-	uid, _ := uuid.NewUUID()
+func newBurstJob(virtClient kubecli.KubevirtClient, workload *config.Workload, uuid string, d <-chan time.Time) *BurstJob {
 	return &BurstJob{
 		virtClient: virtClient,
 		Workload:   workload,
-		UUID:       uid.String(),
+		UUID:       uuid,
 		done:       d,
+		watchers:   map[string]*watcher.ObjListWatcher{},
 	}
 }
 
 func (b *BurstLoadGenerator) Delete(virtClient kubecli.KubevirtClient, workload *config.Workload) {
-	j := newBurstJob(virtClient, workload, b.Done)
-	j.DeleteWorkload()
+	j := newBurstJob(virtClient, workload, b.UUID, b.Done)
+	j.DeleteWorkloads()
+	j.stopAllWatchers()
 }
 
 func (b *BurstLoadGenerator) Run(virtClient kubecli.KubevirtClient, workload *config.Workload) {
-	j := newBurstJob(virtClient, workload, b.Done)
-	j.CreateWorkload()
+	j := newBurstJob(virtClient, workload, b.UUID, b.Done)
+	j.CreateWorkloads()
+	// only stop watchers if the test will not delete the objs, otherwise the watchers will watch for the obj deletions
+	if !flags.Delete {
+		j.stopAllWatchers()
+	}
 }
 
-func (b *BurstJob) CreateWorkload() {
-	log.Log.V(1).Infof("Burst Load Generator CreateWorkload")
-
-	var wg sync.WaitGroup
-	obj := b.Workload.Object
-	for replica := 1; replica <= b.Workload.Count; replica++ {
-		select {
-		case <-b.done:
-			log.Log.V(1).Infof("Burst Load Generator duration has timed out")
-			return
-		default:
-		}
-
-		log.Log.V(2).Infof("Replica %d of %d", replica, b.Workload.Count)
-		newObject, err := utils.Create(b.virtClient, replica, obj, b.UUID)
+func (b *BurstJob) CreateWorkloads() {
+	log.Log.V(1).Infof("Burst Load Generator CreateWorkloads")
+
+	// The watcher must be created before to be able to watch all events related to the objs.
+	// This is important because before creating each obj we need to create a obj watcher for each obj type.
+	objSpec := b.Workload.Object
+	objSample := renderObjSpecTemplate(objSpec, b.UUID)
+	b.createWatcherIfNotExist(objSample)
+	objUtil.CreateNamespaceIfNotExist(b.virtClient, objSample.GetNamespace(), config.WorkloadUUIDLabel, b.UUID)
+
+	// Create all replicas
+	for r := 1; r <= b.Workload.Count; r++ {
+		log.Log.V(2).Infof("Replica %d of %d", r, b.Workload.Count)
+		idx := r
+		_, err := objUtil.CreateObjectReplica(b.virtClient, objSpec, &idx, b.UUID)
 		if err != nil {
 			continue
 		}
-		if b.objType == "" {
-			b.objType = objUtil.GetObjectResource(newObject)
-		}
+	}
 
-		wg.Add(1)
-		go func(newObject *unstructured.Unstructured) {
-			defer wg.Done()
-			b.watchCreate(newObject)
-			log.Log.Infof("obj %s is available", newObject.GroupVersionKind().Kind)
-		}(newObject)
+	// Wait all objects be Running
+	for objType, objWatcher := range b.watchers {
```

Comment:
I think the assumption is that all VMs should be in running state anyway, thus I think the order is irrelevant here. Am I wrong about this? Or do you mean that there might be a kind of wait starvation, in the kind of, "lots of other VMs are coming up before the one I am waiting for currently"?

Reply:
I thought `objWatcher.WaitRunning(b.Workload.Timeout.Duration)` is being called for each object.  My concern is that we're waiting for one object to be completed before we start the timeout for the next one.  I might be getting something confused... because why the for loop...?

Reply:
Actually we have a watcher for each type of object. So the loop exist to wait all obj for each types.

Having the logic to create a watcher for each obj was one of the main thing that I fixed. A watcher will watch all obj for a given type...

In the previous version, it was possible to create different types of object at the same time, for example mix Pods with VMs, or VM and VMI, etc....